[MRG] GSoC 2014: Standard Extreme Learning Machines

IssamLaradji opened the pull request:
Finished implementing the standard extreme learning machines (ELMs). I am getting the following results with 550 hidden neurons against the digits datasets,

Training accuracy using the logistic activation function: 0.999444
Training accuracy using the tanh activation function: 1.000000

Fortunately, this algorithm is much easier to implement and debug than multi-layer perceptron :).
I will push a test file soon.

@ogrisel , @larsmans

IssamLaradji committed:
ENH Extreme Learning Machines for classification and regression

IssamLaradji committed:
fixed the "centering" issue for the figures generated by elm_scale_....py

IssamLaradji commented:
fixed! and pushed the changes 
![centered](https://cloud.githubusercontent.com/assets/3382128/5608064/e697c554-9429-11e4-8477-4dea9eff6b9f.png)

IssamLaradji commented:
Hi @ogrisel , do you think this PR is ready for its final review ? I believe I addressed the mentioned requirements.

Macbull commented:
Hi @IssamLaradji, its a great work and will make it easy for many to use ELM. I was wondering if we can access the output weights of hidden layer i.e. ß?

IssamLaradji commented:
Hi @Macbull,

After training the `elm`, the beta weights are stored in the attribute `.coef_output_`. So you can access it by calling, `elm.coef_output_`.

cheers.

andaag commented:
Ping @ogrisel ref comment from @IssamLaradji 

I've been playing with this branch for a few test datasets and I'd love to see this merged :)

ddofer commented:
Has this been merged into the bleeding edge?

amueller commented:
not yet, sorry.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+
+.. topic:: References:
+
+ *  Huang, Guang-Bin, Qin-Yu Zhu, and Chee-Kheong Siew. 
```
the author format is inconsistent. I guess it should be Guang-Bin Huang?

amueller replied:
I am not sure if this is our place but maybe we should warn that these are non-standard journals and the math is questionable?
Maybe something "These articles were published outside of the machine learning community, and might not have received the same amount of scrutiny as other sources"? Is that to harsh for :hankey:?

IssamLaradji replied:
Haha, I'm not sure. But we could add a link that leads to the journal's description ?or we could add a small description of the journal implicitly describing it as a non-machine learning journal?

Also, we could avoid explaining any mathematical properties of ELM in the document, so we wouldn't have anything to criticize?

I didn't dig much into its theory, but its practical applications is quite astonishing.

amueller replied:
Could you please also add the paper ``Extreme Learning Machine for Regression and
Multiclass Classification``?
For the warning, I'd like other core devs feedback on this.

amueller commented:
could you please rebase and add the model to ``bench_mnist``?

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+     dimensional space;
+  *  the result passes through an element-wise non-linear activation function,
+     typically a Rectified Linear Unit or a sigmoid function (tanh or logistic); and
+  *  last, it trains a linear one vs. rest classifier or a multi-output ridge 
```
I would say "finally" instead of "last".

amueller replied:
I would say "it fits a supervised linear ridge-model, that is a least square model with linear penalty, for both regression and classification". At least that seems to be what is implemented. Your current formulation leaves it open what loss  and penalty are for the classification case.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+  *  ``n_hidden``: controls the number of hidden layer nodes; and
+  *  ``C``: controls the regularization strength of the output linear model.
+
+``weight_scale`` and ``C`` are similar in that they regularize weight parameters
```
Is that not true for all of them? A large ``n_hidden`` will also increase complexity.

amueller commented:
I am just trying to read "Extreme learning machines: Theory and applications". Is it just me or is the proof of Theorem 2.1 just nonsense?

kastnerkyle commented:
I thought it was nonsense as well.

On Thu, Apr 30, 2015 at 3:50 PM, Andreas Mueller <notifications@github.com>
wrote:

> I am just trying to read "Extreme learning machines: Theory and
> applications". Is it just me or is the proof of Theorem 2.1 just nonsense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3306#issuecomment-97947224>
> .
>

amueller commented:
Thanks @kastnerkyle, I thought I was going mad. Haven't done math in a while.

I believe the theorem somewhat. I mean if g is constant zero, this is clearly not true. And last time I checked 0 was pretty differentiable. I'm not sure what the correct conditions on ``g`` are but the whole thing seems paper generated. How can a *single vector* not lie in an N-1 dim subspace? I would think it lies in infinitely many one as long as N > 1.

amueller commented:
Wait, and Theorem 2.2 is "any positive epsilon is greater than zero"?

kastnerkyle commented:
Yeah. The clincher is that they don't even cite Johnson-Lindenstrauss... I
have real problems with this paper.

On Thu, Apr 30, 2015 at 4:01 PM, Andreas Mueller <notifications@github.com>
wrote:

> Wait, and Theorem 2.2 is "any positive epsilon is greater than zero"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3306#issuecomment-97951136>
> .
>

amueller commented:
I'll try for a bit more, but I think I should just review the code. They have so many citations! And ELM is only ever published in neurocomputing and neural networks, which are really questionable....

kastnerkyle commented:
10/26 citations are also self citations! Anyways I will stop. It would be
good if we can find a secondary reference that is a little less insane, but
I don't know if there are any.

I like ELM in certain applications but there are circles where the ELM hype
is even worse than deep learning hype. I have had people tell me that ELMs
killed neural nets... not that I am biased or anything.

My recommendation is to not read the paper too hard and just think of it as
a logical application of JL lemma + linear model. I *really* don't like the
paper :D

On Thu, Apr 30, 2015 at 4:03 PM, Kyle Kastner <kastnerkyle@gmail.com> wrote:

> Yeah. The clincher is that they don't even cite Johnson-Lindenstrauss... I
> have real problems with this paper.
>
> On Thu, Apr 30, 2015 at 4:01 PM, Andreas Mueller <notifications@github.com
> > wrote:
>
>> Wait, and Theorem 2.2 is "any positive epsilon is greater than zero"?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/3306#issuecomment-97951136>
>> .
>>
>
>

amueller commented:
:rage1: :rage2: :rage3: :rage4:

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+.. figure:: ../auto_examples/neural_networks/images/plot_elm_diagram_001.png
+   :target: ../auto_examples/neural_networks/plot_elm_diagram.html
+   :align: center
+   :scale: 100%
```
I would make it smaller. I feel it is way to larger. maybe 50% or less?

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+below:
+
+.. figure:: ../auto_examples/neural_networks/images/plot_elm_scale_C_and_weight_scale_001.png
+   :target: ../auto_examples/neural_networks/plot_elm_scale_C_and_weight_scale.html
```
This seems very streched. I think you should adjust the aspect ratio of the plot.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+
+   * :ref:`example_neural_networks_plot_elm_scale_C_and_weight_scale.py`
+
+Higher ``n_hidden`` provides ELM with a higher capacity to learn complex 
```
Ah, I feel this part belongs together with the part above.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+functions. It allows ELM to have more feature detectors in the hidden 
+layer that characterize the training dataset. 
+
+This implements :ref:`classification <elm_classification>` for classification 
```
I'm not sure what "this" refers to here.

amueller replied:
Maybe just "There are implementations for regression and classification that support both dense ..."?

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+==============
+
+:class:`ELMClassifier` supports binary-class, multi-class, and
+multi-label classification on datasets.
```
I feel "on datasets" is redundant.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+multi-label classification on datasets.
+
+Like other classifiers, :class:`ELMClassifier` trains on a dataset using the
+fit method that accepts two input parameters: an array X of size 
```
possibly backticks around X and y, but no strong opinion.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+Like other classifiers, :class:`ELMClassifier` trains on a dataset using the
+fit method that accepts two input parameters: an array X of size 
+``(n_samples, n_features)``, which represents the training samples, and an array y 
+of integer values of size ``(n_samples)``, which holds the class labels respective
```
"(for the multi-class case)". I think we can leave out the multi-label case in this description but I think we should specify this is the shape for multi-class.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+-----------------------
+
+For imbalanced datasets, giving more weight to classes having less samples than 
+others could help improve accuracy. 
```
I am not sure if "accuracy" is right. Also the example below should probably report some scores, like f1 or auc?

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+in that the final output is a continuous value.
+
+The fit method in :class:`ELMRegressor` accepts arguments X and y
+where y is expected to be a matrix of floating point values::
```
Not a matrix! a vector or for the multi-output case a numpy array.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,222 @@
+.. figure:: ../auto_examples/neural_networks/images/plot_weighted_elm_001.png
+   :target: ../auto_examples/neural_networks/plot_weighted_elm.html
+   :align: center
+   :scale: 100%
```
FYI this dataset here has been expertly crafted for sample weights: http://scikit-learn.org/stable/auto_examples/svm/plot_weighted_samples.html but feel free to keep this one.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+
+  * **Setting batch_size**: training on large datasets or having a large 
+    ``n_hidden`` value could cause issues for limited memory. In this case, it is 
+    recommended to set ``batch_size=1000`` in order to process the data in chunks. 
```
Maybe you should say here that by default, all samples are processed as a whole. And recommend setting "an upper limit such as ``batch_size=1000``".

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+    increase training time.
+
+  * **Preprocessing**: Making sure that the input features vary on the same scale is 
+    important for having consistent prediction results. ``StandardScaler`` and ``**MinMaxScaler``
```
Remove the double stars at MinMaxScaler, and use the ``:class:`` directive to get links.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+    important for having consistent prediction results. ``StandardScaler`` and ``**MinMaxScaler``
+    are two appropriate scalers for this task (see :ref:`link <preprocessing_scaler>`). 
+    Also, a good practice is to use ``Pipeline`` as demonstrated in 
+    :ref:`plot_elm_hyperparameters.py <example_neural_networks_plot_elm_hyperparameters.py>`. 
```
Why do you name this link instead of the automatic naming?

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+on testing score.
+
+.. figure:: ../auto_examples/neural_networks/images/plot_elm_hyperparameters_001.png
+   :target: ../auto_examples/neural_networks/plot_elm_hyperparameters.html
```
Maybe the caption should say which dataset this is, or at least whether this is a regression or classification task, and what the colors mean (accuracy?)

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+uniform distribution; matrix :math:`\beta` contains unknown variables;
+and :math:`\sigma` is the non-linear, component-wise activation function.
+
+ELM solves for :math:`\beta` using the ridge regression implementation, given
```
I think it should say "using ridge regression". The word implementation seems out-of-place in the math section.

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+Related Models
+==============
+
+  * :ref:`Kernel approximation <kernel_approximation>` : this introduces other models
```
Maybe say explicitly, RBFSampler (Random kitchen sinks) can be seen as a random neural net with a cosine non-linearity and sampling weights using a gaussian distribution.

amueller commented on examples/neural_networks/plot_elm_diagram.py:
```diff
@@ -0,0 +1,86 @@
+Plot ELM diagram
+================
+
+This example plots a single-hidden layer feedforward network. It also allows
```
I would not put a dash between single and hidden. Maybe "single-hidden-layer"  or just "single hidden layer"

amueller commented on examples/neural_networks/plot_elm_diagram.py:
```diff
@@ -0,0 +1,86 @@
+"""
+================
+Plot ELM diagram
```
This is really not related to ELM that much, right? Maybe by the fact that the output layer has no bias?

amueller commented on doc/modules/neural_networks_supervised.rst:
```diff
@@ -0,0 +1,255 @@
+.. figure:: ../auto_examples/neural_networks/images/plot_elm_hyperparameters_001.png
+   :target: ../auto_examples/neural_networks/plot_elm_hyperparameters.html
+   :align: center
+   :scale: 100%
```
I would also make this smaller.

amueller commented:
Does this ever do better than RBFSampler? I guess they are kinda the same?

amueller commented on examples/neural_networks/plot_elm_hyperparameters.py:
```diff
@@ -0,0 +1,80 @@
+===========================================================
+
+This example first demonstrates how sklearn.grid_search.GridSearchCV can
+scan through the space of 2 significant hyperparameters C and weight_scale,
```
I would use "important" not "significant" her and in the next sentence. "significant" has a very specific meaning in stats. There is not really a danger of confusion here, still I feel "significant" a bit odd.

amueller commented on examples/neural_networks/plot_elm_hyperparameters.py:
```diff
@@ -0,0 +1,80 @@
+scan through the space of 2 significant hyperparameters C and weight_scale,
+recording their cross-validation scores on the digits dataset.
+Please note that n_hidden is also a significant parameter but was not
+demonstrated here because of the involved computational costs.
```
I would just say "for this demonstration it was set to 1024 to save run-time"

amueller replied:
I would argue that his example runs too long to be a "plot" example (20 minutes).

amueller commented on examples/neural_networks/plot_elm_scale_C_and_weight_scale.py:
```diff
@@ -0,0 +1,138 @@
+Impact of ELMs hyperparameters C and weight_scale on toy datasets
+=================================================================
+
+This illustrates how varying the regularization terms weight_scale, and C can
```
"This example" ?

amueller commented on examples/neural_networks/plot_elm_scale_C_and_weight_scale.py:
```diff
@@ -0,0 +1,138 @@
+underfitting. But it is important not to increase their values so much that
+they lead to high variance which indicates overfitting. Therefore, it is best
+to find a balance between bias and variance (or a value that is not too high
+nor too low) by testing a range of values as seen in this example.
```
Maybe refer to the grid-search example, too?

amueller commented on examples/neural_networks/plot_elm_scale_C_and_weight_scale.py:
```diff
@@ -0,0 +1,138 @@
+                               random_state=rng,
+                               n_clusters_per_class=1)
+
+    X += 2 * rng.uniform(size=X.shape)
```
why not use the ``noise`` paramter of ``make_classification``?

amueller commented on examples/neural_networks/plot_weighted_elm.py:
```diff
@@ -0,0 +1,71 @@
+"""
+==================================
+Weighted Extreme Learning Machines
```
I would call this "class weights in ELM" as in contrast to sample weights

amueller commented on examples/neural_networks/plot_weighted_elm.py:
```diff
@@ -0,0 +1,71 @@
+Plot decision functions of extreme learning machines with different class
+weights. Assigning larger weight to a class will push the decision function
+away from that class to have more of its samples correctly classified.
+Such scheme is useful for imbalanced data so that underrepresented classes
```
"Such a scheme"

amueller commented on examples/neural_networks/plot_weighted_elm.py:
```diff
@@ -0,0 +1,71 @@
+n_hidden = 100
+
+activation = 'tanh'
+clf_weightless = ELMClassifier(n_hidden=n_hidden, activation=activation,
```
maybe a for-loop over the different ``class_weight`` settings?

amueller commented on sklearn/ensemble/tests/test_weight_boosting.py:
```diff
@@ -272,6 +272,23 @@ def test_sample_weight_missing():
     assert_raises(ValueError, clf.fit, X, y_regr)
 
 
+def test_sample_weight_missing():
```
This file doesn't seem to belong in this PR.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+from ..utils.class_weight import compute_sample_weight
+
+
+def _multiply_weights(X, sample_weight):
```
maybe ``_multiply_sample_weights``?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        self.random_state = random_state
+
+    def _init_weights(self, n_features):
+        """Initialize the parameter weights."""
```
"Sample random weights for hidden layer"?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        """Initialize the parameter weights."""
+        rng = check_random_state(self.random_state)
+
+        # Use the initialization method recommended by Glorot et al.
```
This should be documented in the docstring / narrative. Also, is this good for ELM?

kastnerkyle replied:
I don't think there is any result on initialization for ELM.

On Thu, Apr 30, 2015 at 5:13 PM, Andreas Mueller <notifications@github.com>
wrote:

> In sklearn/neural_network/extreme_learning_machines.py
> <https://github.com/scikit-learn/scikit-learn/pull/3306#discussion_r29470879>
> :
>
> > +                 random_state):
> > +        self.C = C
> > +        self.activation = activation
> > +        self.class_weight = class_weight
> > +        self.weight_scale = weight_scale
> > +        self.batch_size = batch_size
> > +        self.n_hidden = n_hidden
> > +        self.verbose = verbose
> > +        self.warm_start = warm_start
> > +        self.random_state = random_state
> > +
> > +    def _init_weights(self, n_features):
> > +        """Initialize the parameter weights."""
> > +        rng = check_random_state(self.random_state)
> > +
> > +        # Use the initialization method recommended by Glorot et al.
>
> This should be documented in the docstring / narrative. Also, is this good
> for ELM?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3306/files#r29470879>.
>

amueller replied:
Well, there have been results on random nets that are not called ELM, right? Not sure about initialization. If we don't know if it helps, it seems like an additional layer of unexpected interference between the scale and n_hidden, right? Maybe we should plot these against each other?

kastnerkyle replied:
Yes, recently the work of A. Saxe has explored this as well as past stuff on reservoir networks and echo state networks. But I have not seen clear evidence that certain random initializations (which are *not* followed by learning) are better or worse than others. The best bet IMO would be to use the same initialization as we use for random_projection to be consistent with ourselves, but if this works too then it probably isn't a huge issue. Maybe there are recent ELM papers which discuss this.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        self.intercept_hidden_ = rng.uniform(-weight_init_bound,
+                                             weight_init_bound,
+                                             self.n_hidden)
+        if self.weight_scale != 1:
```
why not multiply the weight_init_bound? that would save a maxtrix-scaler multiplication?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+        # Apply the activation method
+        activation = ACTIVATIONS[self.activation]
+        hidden_activations = activation(hidden_activations)
```
maybe return in this line?

mblondel replied:
I wanted to do the same comment. Directly return here.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+                                                     ACTIVATIONS))
+
+        # Initialize public attributes
+        if not hasattr(self, 'classes_'):
```
Why is this necessary?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+            self.coef_hidden_ = None
+
+        # Initialize private attributes
+        if not hasattr(self, '_HT_H_accumulated'):
```
Why is this necessary?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+            y = column_or_1d(y, warn=True)
+
+        # Classification
+        if isinstance(self, ClassifierMixin):
```
This looks ugly. There are separate classes, right? Why not add a member ``encode_y`` that is identity for regression and does this in classification? This shouldn't live in the general base-class. I think there is a function like that in GBRT, maybe we should give the same name here.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        self.n_outputs_ = y.shape[1]
+
+        # Step (1/2): Compute the hidden layer coefficients
+        if (self.coef_hidden_ is None or (not incremental and
```
why not just ``hasattr(self.coef_hidden_)`` here?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+            # beta_{0} = inv(H_{0}^T H_{0} + (1. / C) * I) * H_{0}.T y_{0}
+            self.coef_output_ = ridge_regression(H_batch, y[batch_slice],
+                                                 1. / self.C,
```
This might be controversial, but shouldn't C include ``n_samples`` to be comparable to ``SVC``? Or we should use ``alpha`` instead? Which paper is the ``C`` from?

IssamLaradji replied:
This paper has it as `1/C`: http://www.ntu.edu.sg/home/egbhuang/pdf/ELM-Unified-Learning.pdf

I honestly prefer `alpha` because then I could set it to zero if I don't want any regularization.
What do you think ?

amueller replied:
I think you currently don't implement the paper. Alpha is n_samples / C in ridge_regression,  I think. The paper uses the same convention as our SVC, but I think your implementation does not. Can someone confirm? @agramfort @GaelVaroquaux @ogrisel @mblondel ?

amueller replied:
And I don't have a preference regarding C vs alpha.

amueller replied:
I was going by https://github.com/scikit-learn/scikit-learn/issues/759#issuecomment-5063190 but maybe that is wrong? see #4666. So maybe the C is correct.

amueller replied:
yeah seems correct. I was confused by the fact that alpha means different things in ridge vs sparse models.

amueller commented:
Btw @IssamLaradji do you want to work on these comments or would you rather have someone else pick this up?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        batches = gen_batches(n_samples, batch_size)
+
+        # (First time call) Run the least-square algorithm on batch 0
+        if not incremental or self._HT_H_accumulated is None:
```
There seems a lot of repetition between what happens in the first batch and the other batches. Have you tried to refactor that?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+            # Update hidden-to-output coefficients
+            regularized_HT_H = self._HT_H_accumulated.copy()
+            regularized_HT_H.flat[::self.n_hidden + 1] += 1. / self.C
```
same remark about C here.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            The input data.
+
+        y : array-like, shape (n_samples,)
```
The shape of y is wrong for multi-output regression. Maybe ``fit`` should be in the inheriting classes.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Subset of training data.
+
+        y : array-like, shape (n_samples,)
```
again, shape is wrong for multi-output regression.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+            Target values.
+
+        sample_weight : array-like, shape (n_samples,)
+            Per-sample weights. Rescale C per sample. Higher weights
```
"Rescale**s** C per sample". Did you copy & paste this from somewhere?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+            Subset of target values.
+
+        sample_weight : array-like, shape (n_samples,)
+            Per-sample weights. Rescale C per sample. Higher weights
```
Rescales

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+        return self
+
+    def _decision_scores(self, X):
```
why not ``_decision_function``?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+class ELMClassifier(BaseELM, ClassifierMixin):
+    """Extreme learning machine classifier.
+
+    The algorithm trains a single-hidden layer feedforward network by computing
```
either no dash or two dashes as above.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+    """Extreme learning machine classifier.
+
+    The algorithm trains a single-hidden layer feedforward network by computing
+    the hidden layer values using randomized parameters, then solving
```
random weights?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+    The algorithm trains a single-hidden layer feedforward network by computing
+    the hidden layer values using randomized parameters, then solving
+    for the output weights using least-square solutions. For prediction,
```
"using l2 regularized least-squares"?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+    the hidden layer values using randomized parameters, then solving
+    for the output weights using least-square solutions. For prediction,
+    after computing the forward pass, the continuous output values pass
+    through a gate function converting them to integers that represent classes.
```
Maybe rather say thresholded? at zero?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+    weight_scale : float, default 1.
+        Initializes and scales the input-to-hidden weights.
+        The weight values will range between plus and minus
+        'sqrt(weight_scale * 6. / (n_features + n_hidden))' based on the
```
ah it is documented, nice :) still it would be interesting to know if that increases or decreases interactions between n_hidden and weight_scale.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+    batch_size : int, optional, default None
+        If None is given, batch_size is set as the number of samples.
+        Otherwise, it will be set as the given integer.
```
This sentence doesn't contain much information. Maybe rather "Otherwise, the data will be split into chunks of size batch_size, which might increase the computational cost, but limits the memory requirements."

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        If None is given, batch_size is set as the number of samples.
+        Otherwise, it will be set as the given integer.
+
+    verbose : bool, optional, default False
```
I prefer verbose to be integer.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        "Weighted extreme learning machine for imbalance learning."
+        Neurocomputing 101 (2013): 229-242.
+
+    Glorot, Xavier, and Yoshua Bengio. "Understanding the difficulty of
```
Again weird names. Lastname, Firstname and Firstname Lastname.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+            in subsequent calls.
+
+        sample_weight : array-like, shape (n_samples,)
+            Per-sample weights. Rescale C per sample. Higher weights
```
"Rescale**s** C"

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+        Returns
+        -------
+        self : returns a trained elm usable for prediction.
```
capitalize ELM as before?

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,625 @@
+
+    def predict_proba(self, X):
+        """Probability estimates.
+
```
It is now :)

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        y_prob : array-like, shape (n_samples, n_classes)
+            The predicted probability of the sample for each class in the
+            model, where classes are ordered as they are in
+            `self.classes_`.
```
double ticks

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+    The algorithm trains a single-hidden layer feedforward network by computing
+    the hidden layer values using randomized parameters, then solving
+    for the output weights using least-square solutions. For prediction,
```
l2 regularized least-squares regression.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+    The algorithm trains a single-hidden layer feedforward network by computing
+    the hidden layer values using randomized parameters, then solving
+    for the output weights using least-square solutions. For prediction,
+    ELMRegressor computes the forward pass resulting in continuous output
```
I'm not sure what this sentence is supposed to tell me.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+    Parameters
+    ----------
+    C : float, optional, default 100
+        A regularization term that controls the linearity of the decision
```
Is that true? Is that from  the paper? This regularizes a linear model, right? That seems ... odd...

IssamLaradji replied:
Yes, it's based on this equation in this paper (http://www.ntu.edu.sg/home/egbhuang/pdf/ELM-Unified-Learning.pdf),

![capture](https://cloud.githubusercontent.com/assets/3382128/7441958/7d0b3de6-f0b3-11e4-8e4a-1d874adaf763.PNG)

Epsilon is the residual for sample `i`, so it's how much importance you give to the residuals.

It regularizes the non-linear model rather.

amueller replied:
Well, it regularizes the last layer, beta is the coefficients of the last layer, right? So I don't understand why making linear coefficients smaller makes the model more linear.

mblondel replied:
This equation is just a constrained form of the ridge regression objective. LS-SVM = kernel ridge.

mblondel replied:
To be more precise, in LS-SVM, they define a constrained form of the objective then derive a dual in order to introduce kernels. In ridge regression, we usually introduce kernels via the representer theorem but the resulting objective is the same.

amueller replied:
But would you say the regularization in kernel ridge is "how linear the function is"?

mblondel replied:
Indeed, the statement above seems wrong.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+
+    Parameters
+    ----------
+    C : float, optional, default 100
```
Again, not sure about C vs alpha, also, the explanation of C seems odd to me.

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        If None is given, batch_size is set as the number of samples.
+        Otherwise, it will be set as the given integer.
+
+    verbose : bool, optional, default False
```
I prefer an integer

amueller commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+        training deep feedforward neural networks." International Conference
+        on Artificial Intelligence and Statistics. 2010.
+    """
+    def __init__(self, n_hidden=100, activation='relu', weight_scale=1.0,
```
btw, are relus used in any of the references?

amueller commented on sklearn/neural_network/tests/test_elm.py:
```diff
@@ -0,0 +1,393 @@
+from sklearn.utils.testing import assert_not_equal
+
+
+np.seterr(all='warn')
```
why this?

amueller commented on sklearn/neural_network/tests/test_elm.py:
```diff
@@ -0,0 +1,393 @@
+
+
+def test_recursive_and_standard():
+    """Test that recursive lsqr return the same result as standard lsqr."""
```
Does this test something else than the partial_fit test above? Well I guess more tests can't hurt.

amueller commented on sklearn/neural_network/tests/test_elm.py:
```diff
@@ -0,0 +1,393 @@
+
+
+def test_warmstart():
+    """Tests that warm_start reuses past solution only if set to true."""
```
I feel this test should also test that the hidden to output weights are kept / updated. They should give the same results as if presented the whole dataset, right?

amueller commented on sklearn/neural_network/tests/test_elm.py:
```diff
@@ -0,0 +1,393 @@
+
+
+def test_classification():
+    """Test ELMClassifier.
```
All tests should use comments, not docstrings now, so we can more easily track what is happening.

amueller commented on sklearn/utils/class_weight.py:
```diff
@@ -59,3 +60,37 @@ def compute_class_weight(class_weight, classes, y):
     return weight
+
+
+def compute_sample_weight(class_weight, classes, y):
```
There is a function in master. Is this the same?

amueller replied:
The one in master has a different signature, I think, but it can probably be used.

amueller commented:
Apart from some minor refactoring and my nitpicking, this looks good. One thing I'd like discussed is if we want C or alpha. If we have C, I feel it should have the same semantics as in the SVM.
The other thing is that I'm not sure if the current weight scale initialization couples or decouples the n_hidden and scale parameters, and I'd like to see a plot on that.

I'd be happy to help if you don't have time to work on this. And I think this one is much further than the MLP and should be easy to have merged.

IssamLaradji commented:
@amueller thanks for offering your help in this. Can two people work on this the same time ?

I could work on addressing the comments on the code, while someone else works on the document, or vice versa. What do you think ?

amueller commented:
@IssamLaradji My cycles are severely limited, but I can do what I can. If you have time to address any of my comments, that is great!
I'd love to see this merged soonish as I think it is a great contribution. I'll see how much I can pitch in.

GaelVaroquaux commented:
Interesting academic troll in the context of this PR:
http://theanonymousemail.com/view/?msg=ZHEZJ1AJ
We need to be careful about the wordings / references used here.

mblondel commented:
Yann Lecun on ELMs:
https://www.facebook.com/yann.lecun/posts/10152872571572143

amueller commented:
I just wanted to post that @mblondel 
The [reddit thread](http://www.reddit.com/r/MachineLearning/comments/34u0go/yann_lecun_whats_so_great_about_extreme_learning/) has this reference:

http://rduin.nl/papers/icpr_92_random.pdf

amueller commented:
Btw, I commented on the refences above: https://github.com/scikit-learn/scikit-learn/pull/3306#discussion_r29468645

It would be great to get your input on that @GaelVaroquaux @mblondel.

Why do you think this is trolling @GaelVaroquaux ? The math in the 2000+ citation paper is entirely :hanky: I would now actually recommend renaming the estimator to RandomNN or something to not credit the ELM author.

GaelVaroquaux commented:
> Why do you think this is trolling @GaelVaroquaux ? The math in the 2000+
> citation paper is entirely :hanky:

I agree. I think that this is trolling because the tone is extremely
confrontational, but I don't disagree with the conclusions.

> I would now actually recommend renaming the estimator to RandomNN or
> something to not credit the ELM author.

:+1: We should mention the link to extreme learning machines, but not
credit the ELM people.

amueller commented:
yeah we should definitely mention them but also mention prior work. How would you word it if referring to the ELM papers, though?

mblondel commented:
Ideally, we should also make a quick empirical comparison with SVMs and RBF networks. RBF networks can actually be implemented in just a few lines of code in scikit-learn. Something like that (untested):
```python
class RBFNetworkTransformer(BaseEstimator):
    def __init__(self, n_components=30, gamma=1):
        self.n_components = n_components
        self.gamma = gamma
    def fit(self, X):
        self.basis_vectors_ = KMeans(n_clusters=n_components).fit(X).cluster_centers_
        return self
    def transform(self, X):
         return rbf_kernel(X, self.basis_vectors_, gamma=self.gamma)
RBFNetworkClassifier = make_pipeline(RBFNetwork(), RidgeClassifier())
```
An option can also be added to randomly select training points instead of using centroids.

IssamLaradji commented:
Do you think we should change the name ? I mean, it is basically this (http://homepage.tudelft.nl/a9p19/papers/icpr_92_random.pdf)
feedforward neural networks with random weights.

GaelVaroquaux commented:
Yes, I think that we should change the name. Maybe something like
RandomNN, or RandomNeuralNet (I'll let more suited people come up with a
better name).

IssamLaradji commented:
sounds good!

mblondel commented:
The name is not the only problem. We also need to decide whether we want to add this at all (c.f. Yann Lecun's comment).

After quickly reading through the code, it seems to me that we could implement something similar as a pipeline:
```python
random_NN = make_pipeline(RandomProjection(), ActivationTransformer(activation="relu"), Ridge())
```
The pipeline would be in user-land code, which solves the name issue.

In the large-scale setting, we can do
```python
random_transformer = make_pipeline(RandomProjection(), ActivationTransformer(activation="relu"))
for batch in batches:
     Xt = random_transformer.transform(X[batch])
     clf.partial_fit(Xt, y[batch])
```
since `random_transformer` is stateless (apart from the rng maybe).

IssamLaradji commented:
Makes sense! but just to confirm your points, you mean that we should

-  create a small class called `random_NN`  that users can import and use out-of-the-box.

-  add batch support update to `ridge_regression` - I believe that's not implemented yet;

- add a `RandomProjection()` function to scikit-learn that can take these parameters:
             - `weight_scale`: sets the range of values for the random sampling.
             - `activation_function`: which could be `identity`, `relu`, `logistic` and so on.

mblondel commented:
No, we would just provide two transformers: one for the random projection and another for the activation layer. It would be up to the user to combine them with Ridge using a pipeline (we would only provide an example).

We already have a random projection module http://scikit-learn.org/stable/modules/random_projection.html but it doesn't use the uniform distribution. However, the distributions used (Gaussian and sparse) are backed by strong theory. On the other hand, I am not sure projecting with the uniform distribution makes any sense. It could be interesting to try use Gaussian random projection instead. There is also the question of the random intercepts. Do they actually improve accuracy?!

If the two transformers don't make sense separetely, we can combine them into one transformer.

For partial_fit, we can use SGDRegressor although it could be nice to add it to Ridge as well if there's a good way to do it. Actually, reading your code again, I couldn't understand the point of your mini-batch procedure. Isn't the matrix `regularized_HT_H` always of shape n_hidden x n_hidden? In that case, solving the system of linear equation should be cheap even if n_samples is large. So there's no point of looping over small batches...

mblondel commented:
To summarize ELM, the idea is to create low-dimensional non-linear features. Then plugging these features to a regressor or classifier leads to efficient learning. In particular, for ridge regression, we only need to solver a small system of linear equations.

The principle is the same for RBF networks except that the features are created differently: use as features the rbf kernels between training points and centroids or randomly selected training points. This can be thought as fixing the "support vectors" ahead of time, which makes learning easier. The SVM objective is more difficult to solve because the support vectors are not known in advance.

GaelVaroquaux commented:
> We also need to decide whether we want to add this at all (c.f. Yann
> Lecun's comment).

The question is: does it work well, yes or no? Yann is well known for
having non nuanced opinions. I wouldn't take his words for granted.

> After quickly reading through the code, it seems to me that we could implement
> something similar as a pipeline:

> random_NN = make_pipeline(RandomProjection(), ActivationTransformer(activation="relu"), Ridge())

> The pipeline would be in user-land code, which solves the name issue.

> In the large-scale setting, we can do

> random_transformer = make_pipeline(RandomProjection(), ActivationTransformer(activation="relu"))
> for batch in batches:
>      Xt = random_transformer.transform(X[batch])
>      clf.partial_fit(Xt, y[batch])

> since random_transformer is stateless (apart from the rng maybe).

Would it work as well? If so we should probably to an example at least,
and maybe still a simple wrapper object, just to make it easy for users,
if it actually works very well. Many of our users do not know how to use
the pipeline (not good, but we just have to face it).

GaelVaroquaux commented:
> There is also the question of the random intercepts. Do they actually
> improve accuracy?!

I would expect that they are crucial. I expect that it's the combination
of linear kernel (not RBF) and random intercept that does the trick. An
RBF kernel would probably not need the random intercept.

mblondel commented:
> I wouldn't take his words for granted.

Which is why I suggested to do a quick comparison with SVMs and RBF nets. Or maybe someone has experience using them?

> Would it work as well?

It would be the same provided that we create a random projection transformer which does the same as

https://github.com/scikit-learn/scikit-learn/pull/3306/files#diff-cc171f4410591cbb16cfe64bcc31841fR70

I can't say if we use standard random projections instead.

IssamLaradji commented:
In my experience, random neural networks or `random_NN` performs really well. It usually performs the same as single-hidden layer perceptron - although it needs much larger number of hidden neurons.
With some tricks such as stacking random-based autoencoders, you could get surprisingly great results with image datasets. I have been using this for a while, and it was quite a success.

I got emails from several people about using this tool and, disregarding the self-citations, the number of citations indicate the high demand of this algorithm.

 I think it would be great if we just keep it and cite the older papers that came up with this idea. 

I will work on showing some comparison results here today.

amueller commented:
@mblondel most of this PR contains "smartly solving the ridge with partial fit".
That wouldn't be possible with a pipeline.

However, it sounds a bit like it should live in the ridge classifier / regressor and not here, and we should work on our pipeline support?

mblondel commented:
@amueller Yes, ideally, the two contributions (random NN and efficient ridge regression) should be decoupled. But still, I don't understand the point of this incremental fitting algorithm. The most expensive operation within the for loop is solving a n_hidden x n_hidden system of linear equations. All the other operations are basically matrix multiplications. How is looping over batches supposed to help scaling w.r.t. n_samples? Or maybe I miss the point...

mblondel commented on sklearn/neural_network/extreme_learning_machines.py:
```diff
@@ -0,0 +1,629 @@
+    if sample_weight is None:
+        return X
+    else:
+        return X * sample_weight[:, np.newaxis]
```
`*` means matrix multiplication when `X` is a sparse matrix.

mblondel commented:
> we should work on our partial-fit pipeline support?

Yes that would indeed be nice, although this particular transformer would be (mostly) stateless, so the user can just call transform in a for loop as showed in my [comment](https://github.com/scikit-learn/scikit-learn/pull/3306#issuecomment-99760329). For the general case, it is not clear whether the transform steps should be fitted against the full data or not.

amueller commented:
The point of the incremental fitting is to do a partial_fit for ridge, which will mean not inverting the whole matrix at once, and in particular not storing it all at once, as ``n_samples x n_hidden`` might be too large to fit in memory.

amueller commented:
For partial fit pipelines, I would only support stateless transformers for the moment. I agree that you can write a pipeline like you did, but the for-loop is actually a bit more complex as you have to fit once in the first iteration. It is not complicated, but having a convenience function would be nice.

mblondel commented:
The matrix that we need to invert for solving the system is n_hidden x n_hidden for all iterations of the loop over the batches. Storing a n_samples x n_hidden shouldn't be a problem if n_hidden < n_features. If n_hidden > n_features, the usual kernel ridge trick can be used.

amueller commented:
It is assumed h_hidden >> n_features, I think.
The kernel trick only works if ``n_samples`` is small, right?
I think this is supposed to work with say, expanding mnist to 10000 features or something. At least that is the idea that I got. Has anyone tried the mnist benchmark with this code?

amueller commented:
On 05/07/2015 08:05 AM, Mathieu Blondel wrote:
>
> I can't say if we use standard random projections instead.
>
our random projections are normal, ELM is uniform. I wouldn't think it 
makes a difference, but we could try.

amueller commented:
It would be good to have an empirical case where the partial fit actually helps. The incremental fitting is what really makes this code non-trivial. If this is helpful, I'd say merge this with renaming / reassignment of credit, and later refactor into Ridge.

If not, maybe just add a transformer and an example?

mblondel commented:
For a stateless transformer, I presume the fit is mostly needed for input checking and setting the rng? The rng could be set in the first call to transform, although this might break your common tests. In any case, you can just call fit once outside of the for loop.

So, indeeded, the incremental fitting is useful in the n_features < n_hidden < n_samples regime. But this is the usual out-of-core learning setting: your features are too big so you need to build them on the fly and call partial fit on small batches. @agramfort had an example using polynomial features in his pydata talk :)

If we really want to go the estimator way (rather than the transformer way), there is actually a more elegant and concise way to solve the problem using conjugate gradient with a LinearOperator. This technique can be used to solve the system of linear equations without ever materializing the transformed features of size n_samples x n_hidden. This should be like 10 lines of code. See https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/ridge.py#L63 for an example of how this works.

amueller commented:
Yeah stateless transformers and the common tests don't work well together, they are currently manually excluded, and that is something we / I need to fix.
The problem with setting the random weights on the first transform is that this would break if people want to use a transformer object on two different datasets. n_features is usually inferred in ``fit`` and if you use another one in ``transform`` that is an error. I guess you could set it in ``transform`` unless it was set in ``fit``, and if you explicitly want to use it on another dataset, you have to call ``fit``. That is slightly magic, though.

amueller commented:
For better ways to solve the problem: well it could be that n_features is small enough that you could fit n_samples x n_features into ram, but not n_samples x n_hidden. Not sure what typical n_features and n_hidden are.

You are the expert in solving linear problems, I am certainly not, so if there are smarter ways to solve this, then we should go for these.
I didn't mentor this GSoC, I just heard multiple times "this just needs a final review"..

mblondel commented:
> For better ways to solve the problem: well it could be that n_features is small enough that you could fit n_samples x n_features into ram, but not n_samples x n_hidden. Not sure what typical n_features and n_hidden are.

Sorry, when I was talking about generating features on the fly, I was referring to the features generated by the random projection + activation transformer. This is the same setting with polynomial features as well: your original features fit in memory but not the combination features obtained by PolynomialFeatures. But the principle is the same even if you start from your raw data, as long as the transformer used is stateless (e.g., FeatureHasher). In all cases, we loop over small batches of data, transform them and call partial_fit.

GaelVaroquaux commented:
> For partial fit pipelines, I would only support stateless transformers for the moment.

+1: let's tackle the simple things first.

mblondel commented:
@amueller: What's the plan? Will you need to detect whether a transformer
is stateless?

amueller commented:
Did I say I have a plan? ;)
Three possible ways?
1) detect (I have no idea how)
2) annotate as stateless
3) via api: stateless transformers don't need to call ``fit``. Then we don't need to call ``fit`` in ``partial_fit``. This would also allow users to provide a statefull transformer fit on a subset of the data.

amueller commented:
Can we decide what to do with this PR first?
@IssamLaradji put a lot of work into it, and it has been sitting around for way to long. If we feel that the algorithm isn't suited for a classifier / regressor class, we should see what we can salvage and add transformers / examples etc.

mblondel commented:
+1 for a transformer on my side. Instead of using a pipeline of two transformers as I initially suggested, we can maybe create just one transformer that does the random projection and applies the activation functions. This should be fairly straightforward to implement. For the examples, showing how to do grid search with a pipeline would be nice. For the name, of the transformer, maybe `RandomActivationTransformer`?

amueller commented:
Do you think the interative ridge regression here has value or are there better ways to ``partial_fit`` ridge regression?

mblondel commented:
The idea of accumulating the n_hidden x n_hidden matrix is nice but this won't scale if n_hidden is large. If we implement a general partial_fit out of this algorithm, this will crash when people try it on high dimensional data like bag of words. We can add it and recommend not to use it when n_features is large. This would still be useful in some settings where n_samples is huge but n_features is reasonably small. For n_features large, I guess one should use SGD's partial_fit.

amueller commented:
Ok. So lets do the transformer? @IssamLaradji do you want to do that?
Or do you think you don't have time?

I'm not sure about ``RandomActivationTransformer``. Maybe ``NonlinearProjection`` though projection kinda means to a lower dim space. NonlinearRandomFeatures? RandomFeatures?

vene commented:
I don't like `RandomFeatures`, it's way too generic. From the name I'd expect it to simply ignore X and return random features. Out of all the names here, it seems to me like `RandomActivation` is the most specific (it best conveys what the object does). (I'd remove the `Transformer` suffix).

jnothman commented:
aside @amueller re stateless transformers: for this purpose, transformers
that depend only on the type or number of columns of the input should also
be acceptable, just to make things tricky!

On 9 May 2015 at 01:43, Vlad Niculae <notifications@github.com> wrote:

>  I don't like RandomFeatures, it's way too generic. From the name I'd
> expect it to simply ignore X and return random features. Out of all the
> names here, it seems to me like RandomActivation is the most specific (it
> best conveys what the object does). (I'd remove the Transformer suffix).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3306#issuecomment-100276392>
> .
>

IssamLaradji commented:
@amueller yah sure! I can do the transformer. 

So I will open a new pull request for this.

Should the file, containing the algorithm, be under the scikit-learn main directory ?
I mean, would it be something like `from sklearn import RandomActivation` ?

Would the parameters be something like,
-`weight_scale`; which sets the range of values for the uniform random sampling.
- `activation_function`; which could be identity, relu, logistic and so on.

PS: I think for ridge regression there is also feature-wise batch support which scales with `n_features` rather than `n_samples`.

mblondel commented:
One possible place would be the pre-processing module.

mblondel commented:
Actually how about putting it in the neural_network module?

IssamLaradji commented:
sounds good.

IssamLaradji commented:
right, it's only used by neural network algorithms as far as I know, so having it in `neural_network` module is better imo.

IssamLaradji commented:
https://github.com/scikit-learn/scikit-learn/pull/4703 this is a rough implementation of the `RandomActivation` algorithm

amueller commented:
How about "RandomBasisFunction" as a name?

ekerazha commented:
The decoupled approach was the same approach of https://github.com/dclambert/Python-ELM
You had a "random_layer" and you could also pipeline it before a Ridge regression.

Moreover, it also included a MELM-GRBF implementation.

mblondel commented:
Thanks, I added the project to
https://github.com/scikit-learn/scikit-learn/wiki/Third-party-projects-and-code-snippets

ProfFan commented:
Just FYI:
Recently Anton Akusok et al. has implemented ELM in python with MAGMA-based acceleration, under the name of hpelm (pypi:https://pypi.python.org/pypi/hpelm).

agramfort commented:
I don't think this will  ever get merged. Closing. Feel free to reopen if you disagree.
